src/crimson/osd/pg.cc

```diff
@@ -433,15 +435,35 @@ void PG::handle_advance_map(
     pgid.pgid,
     &newup, &up_primary,
     &newacting, &acting_primary);
-  peering_state.advance_map(
+  return peering_state.advance_map(
     next_map,
     peering_state.get_osdmap(),
     newup,
     up_primary,
     newacting,
     acting_primary,
-    rctx);
-  osdmap_gate.got_map(next_map->get_epoch());
+    rctx).then([this, next_map] {
+    osdmap_gate.got_map(next_map->get_epoch());
+    return seastar::make_ready_future<>();
+  });
+}
+
+seastar::future<> PG::pre_state_reset(PGLog& pglog)
+{
+  // FIXME: Since rebuilding missing set would involve I/Os,
+  // and in crimson, we don't expect PeeringState state shift
+  // to wait for any I/O, we do the rebuilding here, right
+  // before going to Reset::react(AdvMap). Maybe in the future,
+  // we should modify PeeringState::PeeringListener to make its
+  // method return an indication whether it required I/Os, and
+  // make PeeringState state shift take into account that returned
+  // indication.
+  if (!pglog.get_missing().may_include_deletes
+      && !peering_state.perform_deletes_during_peering()) {
+    return pglog.rebuild_missing_set_with_deletes(
```

Comment:
Ah, the whole thing is here to support rebuild_missing_set_with_deletes.  Is this actually being invoked?  It shouldn't be -- unless I'm mistaken, this call is only here for an upgrade case crimson will never see.

Reply:
From PeeeringState.h:

  bool perform_deletes_during_peering() const {
    return !(get_osdmap()->test_flag(CEPH_OSDMAP_RECOVERY_DELETES));
  }

From OSDMap.cc:

if (inc.new_require_osd_release >= ceph_release_t::unknown) {
    require_osd_release = inc.new_require_osd_release;
    if (require_osd_release >= ceph_release_t::luminous) {
      flags &= ~(CEPH_OSDMAP_LEGACY_REQUIRE_FLAGS);
      flags |= CEPH_OSDMAP_RECOVERY_DELETES;
    }
  }

Looks like CEPH_OSDMAP_RECOVERY_DELETES is always set on new clusters.  Thus, perform_deletes_during_peering is always false.

Thus, PGLog::may_include_deltas is always set during PeeringState::init to true and this shouldn't be called.  If it is, I think there's probably a bug.

Reply:
More generally, we're not supposed to be doing IO during AdvMap.

Reply:
It seems that PeeringState::init is only invoked when creating pgs, when an OSD with pgs in it boots up, PeeringState::init is not called, so the PGLog::may_include_deltas is false in that case in which the rebuild_missing_set_with_deletes is invoked. Perhaps, this is truely when the problem is? Am I right?

Reply:
Maybe we should just unconditionally set PGLog::may_include_deltas to true when loading pgs, is this feasible?

Reply:
Sure, I'd also assert on the osdmap flag though just to be safe.

Reply:
Oh, on load pg it should be read out of the actual encoded missing set, I think.